keps/sig-apps/19-Graduate-CronJob-to-Stable/README.md

```diff
@@ -0,0 +1,503 @@
+  - Add metrics exposing controller throughput, latency etc.
+- Extend CronJob status field
+  - lastSuccessfulTime: tracks the last time the job completed successfully
+  - nextRunTime: tracks the next time the job will be scheduled
+
+## Proposal
+
+### Plan for promoting CronJob to GA
+
+To promote to GA we would create `batch/v1/CronJob` in the [batch/v1 API](https://github.com/kubernetes/kubernetes/blob/b58777eda0f0077eba79d063c1c8f31cd7dbb8b9/staging/src/k8s.io/api/batch/v1/types.go).
+
+We shall have dual implementation of controller (old and new) co-exist, until the new controller reaches GA.
+The old and new implementation of the controller will be toggled through a feature flag:
+- alpha (disabled by default, can be enabled)
+- beta (enabled by default, can be disabled)
+- GA (enabled by default, cannot be disabled)
+
+### Existing controller
+
+The current implementation of the CronJob controller is different than the other workload controllers. GA workload controllers use informers and caches to reduce the load on API server. Whereas the cronjob controller does a periodic poll and sweep of all the objects and acts on them. The CronJob controller has only one worker doing this.
+
+1. syncs all CronJob objects [every 10 seconds](https://github.com/kubernetes/kubernetes/blob/b58777eda0f0077eba79d063c1c8f31cd7dbb8b9/pkg/controller/cronjob/cronjob_controller.go#L98).
+2. Using pager library, gets all Pods and all CronJobs and [processes them one by one](https://github.com/kubernetes/kubernetes/blob/b58777eda0f0077eba79d063c1c8f31cd7dbb8b9/pkg/controller/cronjob/cronjob_controller.go#L136)
+
+This is not a scalable design and ends up loading the API server. Also this does not follow the [recommended guidelines](https://github.com/kubernetes/community/blob/b58777eda0f0077eba79d063c1c8f31cd7dbb8b9/contributors/devel/sig-api-machinery/controllers.md) for building controllers.
+
+### New controller
+
+With this approach we aim to reduce the potential scale issues (e.g. load on API server and memory usage) when using lots of CronJob objects.
+
+#### Informers and Caches
+
+To reduce the need to list all Jobs and CronJobs frequently to reconcile, we propose to use Informers and WorkQueue based architecture. We would be sharing the [same informer cache](https://github.com/kubernetes/kubernetes/blob/b58777eda0f0077eba79d063c1c8f31cd7dbb8b9/cmd/kube-controller-manager/app/controllermanager.go#L450) as the Job controller uses.
+
+We will follow a controller structure similar to existing workload controllers and as [outlined in the guideline](https://github.com/kubernetes/community/blob/b58777eda0f0077eba79d063c1c8f31cd7dbb8b9/contributors/devel/sig-api-machinery/controllers.md#rough-structure).
+
+```golang
+// CronJobController is 2nd generation controller for cronjobs.
+// It is using informers and workqueue to improve its performance over
+// the old controller.
+type CronJobController struct {
+	queue	   			workqueue.RateLimitingInterface
```

Comment:
sorry about the nit picking but `s/workqueue.RateLimitingInterface/workqueue.DelayingInterface`?

Reply:
Good point, updated.